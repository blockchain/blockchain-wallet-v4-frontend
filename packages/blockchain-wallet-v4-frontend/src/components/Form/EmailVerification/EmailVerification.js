import React from 'react'
import styled from 'styled-components'
import { FormattedMessage } from 'react-intl'

import { Button, HeartbeatLoader, Icon, Text } from 'blockchain-info-components'
import media from 'services/ResponsiveService'
import TextBox from '../TextBox'

const Container = styled.div``

const Row = styled.div`
  display: flex;
  flex-direction: row;
`

const InputRow = styled(Row)`
  button {
    border-radius: 0;
  }
  ${media.mobile`
    flex-direction: column;
  `};
  input {
    @media screen and (min-width: 480px) {
      border-right: none;
    }
  }
`

export const LoadingButton = ({ loading, children, ...rest }) => (
  <Button disabled={loading} {...rest}>
    {loading ? (
      <HeartbeatLoader height='20px' width='20px' color='white' />
    ) : (
      children
    )}
  </Button>
)

export const ChangeButton = styled(LoadingButton)`
  margin-right: 16px;
  ${media.mobile`
    margin-right: 0;
    margin-bottom: 16px;
  `};
`

const EmailLabel = styled(Text)`
  margin-bottom: 32px;
  word-break: break-all;
  ${media.mobile`
    margin-bottom: 16px;
  `};
`

const SuccessLabel = styled(Text)`
  margin-bottom: 16px;
  display: flex;
  flex-direction: row;
  align-items: center;
  > :first-child {
    margin-right: 8px;
  }
`

const Verified = ({ email, className }) => (
  <Container className={className}>
    <SuccessLabel color='success'>
      <FormattedMessage
        id='components.EmailVerification.emailverified'
        defaultMessage='Email Verified'
      />
      <Icon name='check' weight='400' size='13px' color='success' />
    </SuccessLabel>
    <EmailLabel weight='400'>{email}</EmailLabel>
  </Container>
)

const EmailSent = ({
  email,
  className,
  sendVerification,
  changeEmail,
  loading
}) => (
  <Container className={className}>
    <Text weight='300'>
      <FormattedMessage
        id='components.EmailVerification.checkyourinbox'
        defaultMessage='Check you inbox. We sent an email to:'
      />
    </Text>
    <EmailLabel weight='400'>{email}</EmailLabel>
    <InputRow>
      <ChangeButton nature='primary' onClick={changeEmail} loading={loading}>
        <FormattedMessage
          id='components.EmailVerification.changeemail'
          defaultMessage='Change Email'
        />
      </ChangeButton>
      <LoadingButton
        nature='primary'
        onClick={sendVerification}
        loading={loading}
      >
        <FormattedMessage
          id='components.EmailVerification.sendemailagain'
          defaultMessage='Send Again'
        />
      </LoadingButton>
    </InputRow>
  </Container>
)

const EmailInput = ({
  input,
  meta,
  updateEmail,
  className,
  errorBottom,
  label,
  loading
}) => (
  <Container className={className}>
    {label && (
      <label htmlFor='email'>
        <FormattedMessage
          id='components.EmailVerification.label'
          defaultMessage='Enter Your Email Address'
        />
      </label>
    )}
    <InputRow>
      <TextBox
        errorBottom={errorBottom}
        placeholder={
          <FormattedMessage
            id='components.EmailVerification.placeholder'
            defaultMessage='Email'
          />
        }
        input={input}
        meta={meta}
      />
      <LoadingButton nature='primary' onClick={updateEmail} loading={loading}>
        <FormattedMessage
          id='components.EmailVerification.sendnow'
          defaultMessage='Send Now'
        />
      </LoadingButton>
    </InputRow>
  </Container>
)

export default class EmailVerification extends React.PureComponent {
  sendVerification = () => {
    this.props.onVerificationSend(this.props.input.value)
  }

  updateEmail = () => {
    this.props.onUpdate(this.props.input.value)
  }

  changeEmail = () => {
    this.props.onEdit()
  }

  render () {
    const { sendVerification, updateEmail, changeEmail } = this
    const {
      input,
      meta,
      errorBottom,
      className,
      verificationSent,
      verified,
      label,
      loading
    } = this.props
    const email = input.value

    if (verified) return <Verified {...{ email, className }} />
    if (verificationSent)
      return (
        <EmailSent
          {...{
            email,
            sendVerification,
            changeEmail,
            className,
            loading
          }}
        />
      )
    return (
      <EmailInput
        {...{
          input,
          meta,
          updateEmail,
          className,
          errorBottom,
          label,
          loading
        }}
      />
    )
  }
}
