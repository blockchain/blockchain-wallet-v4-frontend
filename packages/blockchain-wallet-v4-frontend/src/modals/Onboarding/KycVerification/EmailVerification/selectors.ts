import { lift } from 'ramda'

import { ExtractSuccess } from '@core/types'
import { model, selectors } from 'data'
import { RootState } from 'data/rootReducer'

const { VERIFY_EMAIL_FORM } = model.components.identityVerification

export const getData = (state: RootState) => {
  const formErrors = selectors.form.getFormSyncErrors(VERIFY_EMAIL_FORM)(state)

  const isUserSddEligibleR = selectors.components.buySell.isUserSddEligible(state)

  const email = selectors.core.settings.getEmail(state).getOrElse('')

  return lift((isUserSddEligible: ExtractSuccess<typeof isUserSddEligibleR>) => ({
    email,
    formErrors,
    isUserSddEligible
  }))(isUserSddEligibleR)
}
