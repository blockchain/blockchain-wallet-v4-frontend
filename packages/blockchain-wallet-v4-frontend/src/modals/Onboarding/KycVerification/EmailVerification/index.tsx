import React, { PureComponent } from 'react'
import { connect, ConnectedProps } from 'react-redux'
import { bindActionCreators, Dispatch } from 'redux'

import { actions, model, selectors } from 'data'
import { RootState } from 'data/rootReducer'
import { VerifyEmailFormValuesType } from 'data/types'

import Loading from '../template.loading'
import { getData } from './selectors'
import Success from './template.success'

const { VERIFY_EMAIL_FORM } = model.components.identityVerification

class VerifyEmail extends PureComponent<Props> {
  componentDidMount() {
    const { isEmailVerified, settingsActions } = this.props
    if (!isEmailVerified) {
      settingsActions.fetchSettings()
    }
    this.props.buySellActions.fetchSDDEligibility()
  }

  handleSubmit = () => {
    const { formValues, identityVerificationActions, securityCenterActions, settingsActions } =
      this.props
    if (formValues) {
      identityVerificationActions.updateEmail(formValues.email)
      securityCenterActions.resendVerifyEmail(formValues.email)
      settingsActions.setEmail(formValues.email)
    }
  }

  onResendEmail = (email: string) => {
    const { securityCenterActions } = this.props
    securityCenterActions.resendVerifyEmail(email)
  }

  render() {
    return this.props.data.cata({
      Failure: () => null,
      Loading: () => <Loading />,
      NotAsked: () => <Loading />,
      Success: (val) => (
        <Success
          {...this.props}
          {...val}
          resendEmail={this.onResendEmail}
          onSubmit={this.handleSubmit}
        />
      )
    })
  }
}

const mapStateToProps = (state: RootState) => ({
  data: getData(state),
  formValues: selectors.form.getFormValues(VERIFY_EMAIL_FORM)(state) as
    | VerifyEmailFormValuesType
    | undefined,
  isEmailVerified: selectors.core.settings.getEmailVerified(state).getOrElse(false)
})

const mapDispatchToProps = (dispatch: Dispatch) => ({
  buySellActions: bindActionCreators(actions.components.buySell, dispatch),
  formActions: bindActionCreators(actions.form, dispatch),
  identityVerificationActions: bindActionCreators(
    actions.components.identityVerification,
    dispatch
  ),
  profileActions: bindActionCreators(actions.modules.profile, dispatch),
  securityCenterActions: bindActionCreators(actions.modules.securityCenter, dispatch),
  settingsActions: bindActionCreators(actions.core.settings, dispatch)
})

const connector = connect(mapStateToProps, mapDispatchToProps)

export type OwnProps = {
  handleClose: () => void
}
export type Props = OwnProps & ConnectedProps<typeof connector>
export type LinkDispatchPropsType = ReturnType<typeof mapDispatchToProps>
export default connector(VerifyEmail)
