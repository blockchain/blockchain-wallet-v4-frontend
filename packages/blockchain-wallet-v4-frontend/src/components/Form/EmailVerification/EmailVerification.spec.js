import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

import EmailVerification, {
  ChangeButton,
  LoadingButton
} from './EmailVerification'

const STUB_EMAIL = 'f@ke.mail'

const props = {
  className: 'cn',
  input: { value: STUB_EMAIL },
  onVerificationSend: jest.fn(),
  onEdit: jest.fn(),
  onUpdate: jest.fn()
}

describe('EmailVerification component', () => {
  describe('rendering', () => {
    it('should render verified state correctly', () => {
      const component = shallow(
        <EmailVerification verified {...props} />
      ).dive()
      const tree = toJson(component)
      expect(tree).toMatchSnapshot()
    })

    it('should render verificationSent state correctly', () => {
      const component = shallow(
        <EmailVerification verificationSent {...props} />
      ).dive()
      const tree = toJson(component)
      expect(tree).toMatchSnapshot()
    })

    it('should render verificationSent loading state correctly', () => {
      const component = shallow(
        <EmailVerification verificationSent loading {...props} />
      ).dive()
      const tree = toJson(component)
      expect(tree).toMatchSnapshot()
    })

    it('should render edit state correctly', () => {
      const component = shallow(<EmailVerification {...props} />).dive()
      const tree = toJson(component)
      expect(tree).toMatchSnapshot()
    })

    it('should render edit loading state correctly', () => {
      const component = shallow(<EmailVerification loading {...props} />).dive()
      const tree = toJson(component)
      expect(tree).toMatchSnapshot()
    })

    it('should render edit with label correctly', () => {
      const component = shallow(<EmailVerification label {...props} />).dive()
      const tree = toJson(component)
      expect(tree).toMatchSnapshot()
    })

    it('should render edit loading with label correctly', () => {
      const component = shallow(
        <EmailVerification loading label {...props} />
      ).dive()
      const tree = toJson(component)
      expect(tree).toMatchSnapshot()
    })
  })

  describe('interaction', () => {
    it('should trigger verification send on send click in verificationSent state', () => {
      const component = shallow(
        <EmailVerification verificationSent {...props} />
      ).dive()
      component.find(LoadingButton).simulate('click')
      expect(props.onVerificationSend).toHaveBeenCalledTimes(1)
      expect(props.onVerificationSend).toHaveBeenCalledWith(STUB_EMAIL)
    })

    it('should trigger edit on change email click in verificationSent state', () => {
      const component = shallow(
        <EmailVerification verificationSent {...props} />
      ).dive()
      component.find(ChangeButton).simulate('click')
      expect(props.onEdit).toHaveBeenCalledTimes(1)
    })

    it('should trigger email update on send click in edit state', () => {
      const component = shallow(<EmailVerification {...props} />).dive()
      component.find(LoadingButton).simulate('click')
      expect(props.onUpdate).toHaveBeenCalledTimes(1)
      expect(props.onUpdate).toHaveBeenCalledWith(STUB_EMAIL)
    })
  })
})
