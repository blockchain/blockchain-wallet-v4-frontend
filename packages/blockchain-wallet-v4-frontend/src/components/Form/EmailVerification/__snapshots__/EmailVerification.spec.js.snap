// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmailVerification component rendering should render edit loading state correctly 1`] = `
<EmailVerification__Container
  className="cn"
>
  <EmailVerification__InputRow>
    <TextBox
      input={
        Object {
          "value": "f@ke.mail",
        }
      }
      placeholder={
        <FormattedMessage
          defaultMessage="Email"
          id="components.EmailVerification.placeholder"
          values={Object {}}
        />
      }
    />
    <LoadingButton
      loading={true}
      nature="primary"
      onClick={[Function]}
    >
      <FormattedMessage
        defaultMessage="Send Now"
        id="components.EmailVerification.sendnow"
        values={Object {}}
      />
    </LoadingButton>
  </EmailVerification__InputRow>
</EmailVerification__Container>
`;

exports[`EmailVerification component rendering should render edit loading with label correctly 1`] = `
<EmailVerification__Container
  className="cn"
>
  <label
    htmlFor="email"
  >
    <FormattedMessage
      defaultMessage="Enter Your Email Address"
      id="components.EmailVerification.label"
      values={Object {}}
    />
  </label>
  <EmailVerification__InputRow>
    <TextBox
      input={
        Object {
          "value": "f@ke.mail",
        }
      }
      placeholder={
        <FormattedMessage
          defaultMessage="Email"
          id="components.EmailVerification.placeholder"
          values={Object {}}
        />
      }
    />
    <LoadingButton
      loading={true}
      nature="primary"
      onClick={[Function]}
    >
      <FormattedMessage
        defaultMessage="Send Now"
        id="components.EmailVerification.sendnow"
        values={Object {}}
      />
    </LoadingButton>
  </EmailVerification__InputRow>
</EmailVerification__Container>
`;

exports[`EmailVerification component rendering should render edit state correctly 1`] = `
<EmailVerification__Container
  className="cn"
>
  <EmailVerification__InputRow>
    <TextBox
      input={
        Object {
          "value": "f@ke.mail",
        }
      }
      placeholder={
        <FormattedMessage
          defaultMessage="Email"
          id="components.EmailVerification.placeholder"
          values={Object {}}
        />
      }
    />
    <LoadingButton
      nature="primary"
      onClick={[Function]}
    >
      <FormattedMessage
        defaultMessage="Send Now"
        id="components.EmailVerification.sendnow"
        values={Object {}}
      />
    </LoadingButton>
  </EmailVerification__InputRow>
</EmailVerification__Container>
`;

exports[`EmailVerification component rendering should render edit with label correctly 1`] = `
<EmailVerification__Container
  className="cn"
>
  <label
    htmlFor="email"
  >
    <FormattedMessage
      defaultMessage="Enter Your Email Address"
      id="components.EmailVerification.label"
      values={Object {}}
    />
  </label>
  <EmailVerification__InputRow>
    <TextBox
      input={
        Object {
          "value": "f@ke.mail",
        }
      }
      placeholder={
        <FormattedMessage
          defaultMessage="Email"
          id="components.EmailVerification.placeholder"
          values={Object {}}
        />
      }
    />
    <LoadingButton
      nature="primary"
      onClick={[Function]}
    >
      <FormattedMessage
        defaultMessage="Send Now"
        id="components.EmailVerification.sendnow"
        values={Object {}}
      />
    </LoadingButton>
  </EmailVerification__InputRow>
</EmailVerification__Container>
`;

exports[`EmailVerification component rendering should render verificationSent loading state correctly 1`] = `
<EmailVerification__Container
  className="cn"
>
  <Text
    altFont={false}
    capitalize={false}
    color="gray-5"
    cursor="inherit"
    italic={false}
    opacity={1}
    size="16px"
    uppercase={false}
    weight="300"
  >
    <FormattedMessage
      defaultMessage="Check you inbox. We sent an email to:"
      id="components.EmailVerification.checkyourinbox"
      values={Object {}}
    />
  </Text>
  <EmailVerification__EmailLabel
    weight="400"
  >
    f@ke.mail
  </EmailVerification__EmailLabel>
  <EmailVerification__InputRow>
    <EmailVerification__ChangeButton
      loading={true}
      nature="primary"
      onClick={[Function]}
    >
      <FormattedMessage
        defaultMessage="Change Email"
        id="components.EmailVerification.changeemail"
        values={Object {}}
      />
    </EmailVerification__ChangeButton>
    <LoadingButton
      loading={true}
      nature="primary"
      onClick={[Function]}
    >
      <FormattedMessage
        defaultMessage="Send Again"
        id="components.EmailVerification.sendemailagain"
        values={Object {}}
      />
    </LoadingButton>
  </EmailVerification__InputRow>
</EmailVerification__Container>
`;

exports[`EmailVerification component rendering should render verificationSent state correctly 1`] = `
<EmailVerification__Container
  className="cn"
>
  <Text
    altFont={false}
    capitalize={false}
    color="gray-5"
    cursor="inherit"
    italic={false}
    opacity={1}
    size="16px"
    uppercase={false}
    weight="300"
  >
    <FormattedMessage
      defaultMessage="Check you inbox. We sent an email to:"
      id="components.EmailVerification.checkyourinbox"
      values={Object {}}
    />
  </Text>
  <EmailVerification__EmailLabel
    weight="400"
  >
    f@ke.mail
  </EmailVerification__EmailLabel>
  <EmailVerification__InputRow>
    <EmailVerification__ChangeButton
      nature="primary"
      onClick={[Function]}
    >
      <FormattedMessage
        defaultMessage="Change Email"
        id="components.EmailVerification.changeemail"
        values={Object {}}
      />
    </EmailVerification__ChangeButton>
    <LoadingButton
      nature="primary"
      onClick={[Function]}
    >
      <FormattedMessage
        defaultMessage="Send Again"
        id="components.EmailVerification.sendemailagain"
        values={Object {}}
      />
    </LoadingButton>
  </EmailVerification__InputRow>
</EmailVerification__Container>
`;

exports[`EmailVerification component rendering should render verified state correctly 1`] = `
<EmailVerification__Container
  className="cn"
>
  <EmailVerification__SuccessLabel
    color="success"
  >
    <FormattedMessage
      defaultMessage="Email Verified"
      id="components.EmailVerification.emailverified"
      values={Object {}}
    />
    <Icon
      color="success"
      cursor={false}
      name="check"
      size="13px"
      weight="400"
    />
  </EmailVerification__SuccessLabel>
  <EmailVerification__EmailLabel
    weight="400"
  >
    f@ke.mail
  </EmailVerification__EmailLabel>
</EmailVerification__Container>
`;
